allennlp/modules/token_embedders/embedding.py

```diff
@@ -296,3 +352,183 @@ def _read_pretrained_hdf5_format_embedding_file(embeddings_filename: str, # pyli
                         list(embeddings.shape), [vocab.get_vocab_size(namespace), embedding_dim]))
 
     return torch.FloatTensor(embeddings)
+
+
+def format_embeddings_file_uri(main_file_path_or_url: str,
+                               path_inside_archive: Optional[str] = None) -> str:
+    if path_inside_archive:
+        return "({})#{}".format(main_file_path_or_url, path_inside_archive)
+    return main_file_path_or_url
+
+
+class EmbeddingsFileURI(NamedTuple):
+    main_file_uri: str
+    path_inside_archive: Optional[str] = None
+
+
+def parse_embeddings_file_uri(uri: str) -> 'EmbeddingsFileURI':
+    match = re.fullmatch('\((.*)\)#(.*)', uri)      # pylint: disable=anomalous-backslash-in-string
+    if match:
+        fields = cast(Tuple[str, str], match.groups())
+        return EmbeddingsFileURI(*fields)
+    else:
+        return EmbeddingsFileURI(uri, None)
+
+
+class EmbeddingsTextFile(Iterator[str]):
+    """
+    Utility class for opening embeddings text files. Handles various compression formats,
+    as well as context management.
+
+    Parameters
+    ----------
+    file_uri: str
+        It can be:
+
+        * a file system path or a URL of an eventually compressed text file or a zip/tar archive
+          containing a single file.
+        * URI of the type ``(archive_path_or_url)#file_path_inside_archive`` if the text file
+          is contained in a multi-file archive.
+
+    encoding: str
+    cache_dir: str
+    """
+    DEFAULT_ENCODING = 'utf-8'
+
+    def __init__(self,
+                 file_uri: str,
+                 encoding: str = DEFAULT_ENCODING,
+                 cache_dir: str = None) -> None:
+
+        self._uri = file_uri
+        self._encoding = encoding
+        self._cache_dir = cache_dir
+
+        self._handle: TextIO
+        self._archive_handle: Any = None   # only if the file is inside an archive
+        self._iterator: Iterator[str]
+        self._num_tokens: Optional[int]
```

Comment:
I'm not a fan of these bare declarations, can't you just use the type hints the first time the variable is defined?

Reply:
Sure.

Reply:
👍